### ISSUE TITLE
Query with broken link cannot be cancelled while reading from S3

### ISSUE THREAD

**Issue by yariks5s** [ClickHouse/ClickHouse#58479]
**Describe what's wrong**

When reading data from broken s3 link, the process of reading is stuck for a long time and it's not possible to cancel the query. It's because we check if query is cancelled between processes that were already done, and in this case we cannot check this because the process is never done.

**How to reproduce**

```
SELECT * FROM s3('https://s3queue.clickhouse.com.amazonaws.com/data/*.json', 'JSONEachRow') LIMIT 1
```

**Expected behavior**

The query should be cancelled in the right way without waiting until all retries are done.

**Comment by CheSema** [comment]
`checkTimeLimit` does not work, that means query execution time limit doesn't work.
There are cancelation points only between blocks. But what if calculation of one block is slow.

Client side cancelation doesn't work. 
```
:) set function_sleep_max_microseconds_per_block=100000000
:) select sleepEachRow(1) from system.numbers LIMIT 20

SELECT sleepEachRow(1)
FROM system.numbers
LIMIT 20

Query id: 4cd02e1f-5ead-4aa0-ac84-66603f1ae2cd

Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 20.012 sec.
```
Server does read the `Protocol::Client::Cancel` packet.
I see the logs
```2024.01.03 19:23:37.945059 [ 1944413 ] {1d949579-c544-45f0-b34f-1f5abedc2eed} <Information> TCPHandler: Change cancellation status from NOT_CANCELLED to FULLY_CANCELLED. Log message: Query was cancelled.
```
But it has no effect on result.


```
explain SELECT * FROM s3('https://s3queue.clickhouse.com.amazonaws.com/data/*.json', 'JSONEachRow') LIMIT 1
```
That thing hungs as well. It because that request perform s3 requests actually. It build S3Source which calls listings. And S3Source works as PullingPipelineExecutor, not as Pulling**Async**PipelineExecutor

**Comment by CheSema** [comment]
That approach is good.
https://github.com/ClickHouse/ClickHouse/pull/58314#issuecomment-1875409157

Also S3Source has to be build as PullingAsyncPipelineExecutor.

**Comment by CheSema** [comment]
It seems like function sleep now work with cancel query.
https://github.com/ClickHouse/ClickHouse/pull/57746/files

But still, Cntr^C would stop query only between blocks.
If we set `is_killed=true` when client's cancel packet is read, behaviour would be better.

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Query with broken link cannot be cancelled while reading from S3**

**ClickGap verified:** reproduced it with the script below on `master`; tested 6 version(s) — 6 affected, 0 clean (see Affects); long-standing (not a recent regression); introducing PR not yet pinned.

### Reproduction
```
-- Point s3() at an unreachable/hanging endpoint and ask for a 2-second limit.
-- (The reporter's original host s3queue.clickhouse.com.amazonaws.com no longer
--  resolves; 192.0.2.1 is a guaranteed-hanging address that reproduces the same
--  S3-listing hang.)
SELECT * FROM s3('http://192.0.2.1/data/*.json', 'JSONEachRow') LIMIT 1
SETTINGS max_execution_time = 2, s3_use_adaptive_timeouts = 0,
         s3_retry_attempts = 2, s3_connect_timeout_ms = 1000;
-- Observed: the query IGNORES max_execution_time and runs the full ~9.5s S3
-- retry sequence before failing with Code: 499, instead of aborting at ~2s.
-- Ctrl-C / KILL QUERY are likewise ignored until the listing finishes.
```
[▶ Run on ClickHouse Fiddle](https://fiddle.clickhouse.com/9230af03-a69d-4f2f-a44d-c0ed1068396f)

### Bisect
- Reproduces on every release line tested down to `v25.8.24.21` (the oldest with a downloadable binary for this repro). The introducing change predates `v25.8.24.21`, so it can't be pinned from release binaries — this is a **long-standing** issue, not a recent regression. A maintainer with full git history can `git blame` the relevant file/line to find the exact change.

### Affects
- **Reproduces on**: `master, 26.5, 26.4, 26.3, 26.2, 26.1`

### Suggested next step
Manual triage of the `comp-object-storage` path is needed. The reproducer above should be a fast first check.

**Component:** `comp-object-storage`
**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_58479`_




### Triage metadata
- Primary component: `comp-object-storage`
- Secondary components: `comp-table-functions`

**Comment by groeneai** [comment]
A production report on this family came my way (a replica reached 1000/1000
`max_concurrent_queries` and refused all traffic during a partial S3 degradation), so I
traced the read path on master `484b00ee2610ef5`. Posting the mechanism here because the
remaining question is a policy one I do not think I should answer unilaterally.

### What actually bounds a degraded read

For a data GET the retry stack is:

- inner loop `ReadBufferFromS3.cpp:190-232`, bounded by `max_single_read_retries` (4);
- outer loop is the **AWS SDK's** `AttemptExhaustively`
  (`contrib/aws/.../AWSClient.cpp:300-396`) via our `RetryStrategy`, bounded by
  `s3_retry_attempts` (500). Note it is *not* `Client::with_retries`. `GetObject` uses
  plain `doRequest` (`Client.cpp:531-535`), and that function's own comment (`:782-784`)
  excludes streaming responses.

No wall-time term exists in that stack. The only wall-clock bound is the asynchronous
`CancellationChecker` (`CancellationChecker.cpp:44`), which fires
`cancelQuery(CancelReason::TIMEOUT)`; the stuck thread observes it in
`Client::RetryStrategy::ShouldRetry` (`Client.cpp:104-105`), and for a foreground query
that predicate chain is live (`ThreadStatusExt.cpp:116-122` -> `isCurrentQueryKilled`).

But `appendTask` returns early when the timeout is <= 0 (`CancellationChecker.cpp:83-86`),
so at the default `max_execution_time = 0` the watchdog never arms. Worst case, sleeps only:
4 x sum(min(25*2^n, 5000), n<500) = **~2.7 hours**, and up to ~16.7 h if each request also
burns `s3_request_timeout_ms`. The read holds its slot for that window
(`non_internal_processes`, `ProcessList.cpp:174-182`, decremented only in
`~ProcessListEntry` `:479`), which is what turns a partial degradation into a total refusal.

So the retry loop is cancellable; it just has no deadline of its own, and the one bound
available is off by default. Worth noting the reported `Code: 159` rows can only come from
the arm where `max_execution_time` was set; at the default there is no error, just a held
slot.

A persistently missing object (404) is unaffected: it fails fast via
`isRetryableError` / `processException` (`ReadBufferFromS3.cpp:375`).

### Two gaps in neighbouring paths

- `Client::slowDownAfterRetryableError` (`Client.cpp:910-936`) is an unbounded
  `for (;;) { ... sleepForMilliseconds(...) }` on a Client-shared atomic with no
  cancellation check, while its sibling `ShouldRetry` has one. Only reached from
  `with_retries`, so listing/head/upload, not data GETs.
- `ReadBufferFromAzureBlobStorage` (loops at `:121`, `:254`, `:365`) has no cancellation
  check at all, so on Azure even setting `max_execution_time` does not bound the read.

I have not run either, so I am reporting them rather than claiming impact.

### The policy question

I am not opening a PR, for three reasons:

1. #83957 already occupies this design space. It does not fix the default case (both its
   guards require `max_execution_time > 0`), but it edits exactly these two functions.
2. @ CheSema's review there sets doctrine against the obvious shape: "*retries should know
   about queries settings ... cancelation points are not required to know internal
   mechanics*", preferring one cancellation flag plus an interruptible wait. (FWIW the SDK
   sleep at `AWSClient.cpp:395` is already a `wait_for` on a condition variable in
   `HttpClient::RetryRequestSleep`, but both it and its waker are non-virtual and
   per-HttpClient, i.e. per-disk, so waking it would abort every concurrent read on that
   disk.)
3. Choosing a default deadline is a semantics decision. `max_execution_time = 0` is an
   explicit request for *no* deadline, so an implicit S3-only timeout would override a
   stated intent and fail slow reads that are valid today. And #89051 raised
   `s3_retry_attempts` to 500 deliberately, to survive slow-down responses lasting over ten
   minutes.

One thing that does look tractable: ordinary reads and backups are **already separable**:
`backup_restore_s3_retry_attempts` (1000) is documented as taking effect only for
backup/restore, and `S3BackupClientCreator` runs on a cloned client with its own retry
strategy (`BackupIO_S3.cpp:92-114`). So a deadline could apply to ordinary reads (e.g. at
`S3ObjectStorage::readObject`) without touching the backup guarantee. That still needs a
default value and an owner's call on the semantics.

Is a default read deadline wanted, and if so should it be an S3-read-only budget or should
this stay purely `max_execution_time`-driven? Happy to implement whichever direction is
preferred.
